### PR TITLE
687 timeline rework

### PR DESCRIPTION
**Pull request by Dammshine**

### CODE REVIEW

File: client/src/Services/useSocketCommunication.ts [hunk 1]
```diff
@@ -8,26 +8,35 @@ import {
...
-        console.error(data);
-        setActiveSession(true);
-      },
-      sendFunctionDeclaration: (data: FunctionStructure) => {
-        // TODO: Implement
-        console.error('Received functional structure', data);
+      mainDebug: (_data: 'Finished mainDebug event on server') => {
+        setActive(true);
       },
+      sendFunctionDeclaration: (_data: FunctionStructure) => {},
```
Comment (dqna64) [inline client/src/Services/useSocketCommunication.ts:39]: what is this for

File: client/src/Services/useSocketCommunication.ts [hunk 2]
```diff
@@ -8,26 +8,35 @@ import {
...
   const { socketClient } = useSocketClientStore();
-  const [activeSession, setActiveSession] = useState<boolean>(false);
   const [consoleChunks, setConsoleChunks] = useState<string[]>([]);
   const { updateCurrFocusedTab } = useGlobalStore();
 
-  useEffect(() => {
+  const [, setTaskQueue] = useState<Task[]>([]);
+  const [taskId, setTaskId] = useState(0);
+
+  if (!initialized) {
```
Comment (dqna64) [inline client/src/Services/useSocketCommunication.ts:34]: can use `useRef` or `useMemo` to create this functions only on first render

gpt4o:

```js
import React, { useMemo } from 'react';

const MyComponent = () => {
  const functions = useMemo(() => ({
    func1: () => { /* function logic */ },
    func2: () => { /* function logic */ },
    // Add more functions as needed
  }), []);

  const { func1, func2 } = functions;

  return (
    <div>
      {/* Use your functions here */}
      <button onClick={func1}>Run Function 1</button>
      <button onClick={func2}>Run Function 2</button>
    </div>
  );
};

export default MyComponent;
```

- useRef returns a mutable ref object whose .current property is initialized to the argument passed to useRef (in this case, null).
- The object returned by useRef will persist for the full lifetime of the component.
- By checking if functionsRef.current is null, you can ensure the functions are only created once.

Comment (Dammshine) [reply client/src/Services/useSocketCommunication.ts]: The useMemo hook in React doesn't create a single, always-the-same instance (which is what a "singleton" pattern does). Instead, useMemo is more like a "scoped" function that only runs once per component, but it doesn't guarantee that the instance (in this case, a socket) will be the same every time the component re-renders.

In practice, useSocketCommuication were used in multiple place and thus re-renders in multiple palces.

File: client/src/Services/useSocketCommunication.ts [hunk 3]
```diff
@@ -1,91 +1,169 @@
...
-    return () => {
-      socketClient.clearEventHandlers(handlers);
-    };
-  }, [socketClient]);
+      socketClient.setupEventHandlers(eventHandler);
+    }, []);
+    initialized = true;
+  }
 
   const resetDebugSession = useCallback(() => {
```
Comment (NE1NN) [inline client/src/Services/useSocketCommunication.ts:77]: just wondering why u use so many useCallback/useMemo. Are they all resource-intensive tasks?

File: client/src/components/Button/Button.tsx
Comment (NE1NN) [inline client/src/components/Button/Button.tsx:1]: why didnt u just put the code into index.tsx?

File: client/src/visualiser-debugger/Component/CodeEditor/CodeEditor.tsx [hunk 4]
```diff
@@ -6,12 +6,12 @@ import 'ace-builds/src-noconflict/theme-tomorrow';
...
 import { useUserFsStateStore } from '../../Store/userFsStateStore';
 import { IFileFileNode } from '../FileTree/FS/IFileSystem';
+import { useFrontendStateStore } from '../../Store/frontendStateStore';
 
-type CodeEditorProps = {
-  currLine?: number;
-};
-
-const CodeEditor: React.FC<CodeEditorProps> = ({ currLine = 0 }: CodeEditorProps) => {
+const CodeEditor: React.FC = () => {
```
Comment (NE1NN) [inline client/src/visualiser-debugger/Component/CodeEditor/CodeEditor.tsx:11]: I think React.FC is not needed anymore

https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/function_components/

File: client/src/Services/useSocketCommunication.ts (hunk 3 above)
Comment (Dammshine) [reply client/src/Services/useSocketCommunication.ts:77]: UseCallback and useMemo help prevent these re-renders, it's always good to use no? 

Also, I need the functions below to be updated whenever the dependencies changes.

File: client/src/components/Button/Button.tsx
Comment (Dammshine) [reply client/src/components/Button/Button.tsx:1]: Ya, it's following the convention style of components directory.

**Pull request merged**